sdks/go/pkg/beam/core/util/reflectx/calls.tmpl

```diff
@@ -0,0 +1,68 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reflectx
+
+import "reflect"
+
+// Generated arity-specialized Caller implementations to avoid runtime temporary
```

Comment:
<!--new_thread; commit:5ff88b0ad86fa84eb8323c61651be06916571224; resolved:0-->
It says it's intended to avoid runtime temporary slices but it seems like that's all it's doing in order to call into the Caller interface you've set up. 
Is that what's happening or is this generation just to avoid the reflection overhead?

Reply:
This generation is to create arity-specific caller interfaces, so that using predicates of unknown types can avoid the extra slice allocations. It's also more convenient to use, similarly  to ParDo2, ParDo3, etc. Generated code implements both the general Caller and whatever CallerXxY matches the type. The shims allow an arbitrary general or not-arity-specialized caller to be used as Caller2x1, for example (if types match), so that the user code can always use the more convenient way. 

The filter or combine code thus work whether or not we have a fast type-specialized implementation.